rohmu/object_storage/base.py

```diff
@@ -164,59 +175,63 @@ def get_contents_to_file(self, key, filepath_to_store_to, *, progress_callback:
                 os.unlink(filepath_to_store_to)
             raise
 
-    def get_contents_to_fileobj(self, key, fileobj_to_store_to, *, progress_callback: ProgressProportionCallbackType = None):
+    def get_contents_to_fileobj(
+        self, key: str, fileobj_to_store_to: BinaryIO, *, progress_callback: ProgressProportionCallbackType = None
+    ) -> Metadata:
         """Like `get_contents_to_file()` but writes to an open file-like object."""
         raise NotImplementedError
 
-    def get_contents_to_string(self, key):
+    def get_contents_to_string(self, key: str) -> tuple[bytes, Metadata]:
         """Returns a tuple (content-byte-string, metadata)"""
         with BytesIO() as buf:
             metadata = self.get_contents_to_fileobj(key, buf)
             return buf.getvalue(), metadata
 
-    def get_file_size(self, key):
+    def get_file_size(self, key: str) -> int:
         """Returns an int indicating the size of the file in bytes"""
         # This method isn't currently used by PGHoard itself, it is merely provided
         # for applications that use PGHoard's object storage abstraction layer.
         raise NotImplementedError
 
-    def get_metadata_for_key(self, key):
+    def get_metadata_for_key(self, key: str) -> Metadata:
         raise NotImplementedError
 
-    def list_path(self, key, *, with_metadata=True, deep=False):
+    def list_path(self, key: str, *, with_metadata: bool = True, deep: bool = False) -> list[Any]:
```

Comment:
@fingon  The problem here is that `iter_key` returns a sequence of `IterKeyItem` which is defined as:

```
IterKeyItem = namedtuple("IterKeyItem", ["type", "value"])
```

so the `value` attribute is `Any` . Maybe we should redefine this to be generic in some way so we can use the type variable to provide a better result type? Or can you tell me if the `value` is actually always fixed?

Reply:
Okay, so we could use this definition:

```
from typing import TypeVar, Generic, NamedTuple
Value = TypeVar("Value")


class IterKeyItem(NamedTuple, Generic[Value]):
    type: str
    value: Value
```

But then all uses of `IterKeyItem` will need a type argument . I guess the `Value` type depends on the provider?

Reply:
no, value is str and type is enum (either TYPE_PREFIX or TYPE_OBJECT)

Reply:
@fingon  Are you sure? 

In base.py inside `BaseTransfer.delete_tree` this is what I find:
```
names = [item["name"] for item in self.list_path(key, with_metadata=False, deep=True)]
```

and `list_path` is just a wrapper for `list_iter`:
```
    def list_path(self, key: str, *, with_metadata: bool = True, deep: bool = False) -> list[Any]:
        return list(self.list_iter(key, with_metadata=with_metadata, deep=deep))
```

Which yields the `value`s from `iter_key`:

```
    def list_iter(self, key: str, *, with_metadata: bool = True, deep: bool = False) -> Iterator[Any]:
        for item in self.iter_key(key, with_metadata=with_metadata, deep=deep):
            if item.type == KEY_TYPE_OBJECT:
                yield item.value
```

So `value` can't always be `str`, otherwise the `item["name"]` in delete tree would fail.

Or am I missing something?

Reply:
Ah, right, the value's type depends on the type - PREFIX is str and OBJECT is indeed (platform dependant) dict. Even for those, we should really define fixed TypedDict but perhaps not worth it yet.